runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
+      status = iree_hal_hip_semaphore_query_locked(semaphore, &current_value);
+      if (!iree_status_is_ok(status)) {
+        break;
+      }
+    }
   }
 
-  *out_event = iree_hal_hip_event_handle(wait_timepoint->timepoint.device_wait);
+  if (iree_status_is_ok(status)) {
+    // The current value stored in the semaphore is greater than the current
+    // value, so we can return.
+    if (semaphore->current_visible_value >= value) {
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      iree_hal_hip_event_semaphore_advance(base_semaphore);
+    } else if (iree_timeout_is_infinite(timeout)) {
+      // This is the fast-path. Since we have an infinite timeout, we can
+      // wait directly on the hip event.
+
+      // The current value is not enough, but we have at least submitted
+      // the work that will increment the semaphore to the value we need.
+      // Use iree_hal_hip_util_tree_lower_bound to find the first element in the
+      // tree that would signal our semaphore to at least the given value.
+      iree_hal_hip_util_tree_node_t* node = iree_hal_hip_util_tree_lower_bound(
+          &semaphore->event_queue.tree, value);
+      IREE_ASSERT(
+          node,
+          "We really should either have an event in the queue that will satisfy"
+          "this semaphore, (we checked max_value_to_be_signaled above), or we"
+          "should already have signaled (current_visible_value above)");
+      iree_hal_hip_semaphore_queue_item_t* item =
+          (iree_hal_hip_semaphore_queue_item_t*)
+              iree_hal_hip_util_tree_node_get_value(node);
+
+      iree_hal_hip_event_t* event = item->event;
+
+      // Retain the event, as the event may be removed from the tree
+      // while we sleep on the event.
+      iree_hal_hip_event_retain(event);
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      iree_hal_hip_event_semaphore_advance(base_semaphore);
+      status = IREE_HIP_CALL_TO_STATUS(
+          semaphore->symbols,
+          hipEventSynchronize(iree_hal_hip_event_handle(event)));
+      iree_hal_hip_event_release(event);
+    } else {
+      // If we have a non-infinite timeout, this is the slow-path.
+      // because we will end up having to wait for either the
+      // cleanup thread, or someone else to advance the
+      // semaphore.
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      iree_hal_hip_cpu_event_t* cpu_event = NULL;
+      status = iree_hal_hip_semaphore_get_cpu_event(base_semaphore, value,
+                                                    &cpu_event);
+      if (iree_status_is_ok(status)) {
+        // If there is no cpu event the semaphore has hit the value already.
+        if (cpu_event) {
+          status = iree_wait_one(&cpu_event->event, deadline_ns);
+          iree_hal_resource_release(&cpu_event->resource);
+        }
+      }
+    }
+  }
+
+  // If we are ok status, we expect that the mutex was unlocked.
+  // If there was an error the mutex is still locked.
```

Comment:
Doing this sometimes locked/sometimes not makes the function very difficult to maintain. Prefer clear regions where the lock is held and clear conditions for when it is released - having to trace back through the 4-deep nested ifs to try to find the unlock and make sure it is performed in all cases where status may be ok is not fun. I think I spot a bug with it but it's hard to tell: if iree_hal_hip_semaphore_get_cpu_event or iree_wait_one fails then the lock is not held and the unlock is invalid. same with hipEventSynchronize. I stopped being able to track it after that and it's why this is not a great way to juggle locks :)

if you need to split code into functions that happen under the lock vs those that don't or other refactorings it's worth it - then you can add preconditions to those functions or name them appropriately ("do_foo_with_lock") to make it clear to readers when the lock is held or not and can keep the logic here simpler. deadlocks in error cases are super painful to debug and it's why avoiding requiring locking at all is the best when possible and otherwise keeping them as small/localized as possible.